include/deal.II/lac/solver_bicgstab.h

```diff
@@ -470,45 +403,13 @@ SolverBicgstab<VectorType>::solve(const MatrixType &        A,
 {
   LogStream::Prefix prefix("Bicgstab");
 
-  // Allocate temporary memory.
-  Vr    = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vrbar = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vp    = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vy    = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vz    = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vt    = typename VectorMemory<VectorType>::Pointer(this->memory);
-  Vv    = typename VectorMemory<VectorType>::Pointer(this->memory);
-
-  Vr->reinit(x, true);
-  Vrbar->reinit(x, true);
-  Vp->reinit(x, true);
-  Vy->reinit(x, true);
-  Vz->reinit(x, true);
-  Vt->reinit(x, true);
-  Vv->reinit(x, true);
-
-  Vx = &x;
-  Vb = &b;
-
-  step = 0;
-
   IterationResult state(false, SolverControl::failure, 0, 0);
   do
     {
-      state = iterate(A, preconditioner);
+      state = iterate(A, x, b, preconditioner, state.last_step);
```

Comment:
I actually tried to move everything into `solve`, but then I was discouraged by the various places where the `iterate` function might jump out and request a new attempt with a clean residual and no history, so I gave up. I am sure it can be further restructured, but I realized it would take more time than I wanted to spend (which is probably better spent once we do the bigger task of integrating similar tricks as in the `SolverCG` and need to restructure the code a bit more).

Reply:
I see. Let's keep it like this for now.